src/common/hist_util.cc

```diff
@@ -422,25 +488,49 @@ void GHistIndexMatrix::Init(DMatrix* p_fmat, int max_num_bins) {
       }
     }
 
-    index.resize(row_ptr[rbegin + batch.Size()]);
+    const size_t n_offsets = cut.Ptrs().size() - 1;
+    if ((max_num_bins - 1 <= static_cast<int>(std::numeric_limits<uint8_t>::max())) && isDense) {
+      index.setBinBound(UINT8_BINS_TYPE);
+      index.resize((sizeof(uint8_t)) * row_ptr[rbegin + batch.Size()]);
+    } else if ((max_num_bins - 1 > static_cast<int>(std::numeric_limits<uint8_t>::max())  &&
+      max_num_bins - 1 <= static_cast<int>(std::numeric_limits<uint16_t>::max())) && isDense) {
+      index.setBinBound(UINT16_BINS_TYPE);
+      index.resize((sizeof(uint16_t)) * row_ptr[rbegin + batch.Size()]);
+    } else {
+      index.setBinBound(UINT32_BINS_TYPE);
+      index.resize((sizeof(uint32_t)) * row_ptr[rbegin + batch.Size()]);
+    }
```

Comment:
It can be rewritten by creation of template function again.

Reply:
it was moved to specific func, I hope it's more clear now